<center>
    <img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


In [6]:
%pip install yfinance
#!pip install pandas
#!pip install requests
%pip install bs4
#!pip install plotly

  Using cached frozendict-2.4.1-cp311-cp311-macosx_11_0_arm64.whl
  Using cached peewee-3.17.1-py3-none-any.whl
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.0/73.0 kB 996.9 kB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 112.2/112.2 kB 1.7 MB/s eta 0:00:00 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.5/8.5 MB 4.7 MB/s eta 0:00:00a 0:00:01
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [18]:
%pip install html5lib

Note: you may need to restart the kernel to use updated packages.


In [1]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.default = 'notebook'
import plotly.offline as pyo
pyo.init_notebook_mode(connected=True)

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [2]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [3]:
tesla = yf.Ticker('TSLA')

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [4]:
tesla_data = tesla.history(period = 'max')


TSLA: No price data found, symbol may be delisted (1d 1925-05-01 -> 2024-04-07)


**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [5]:
tesla_data = tesla.history(period = 'max')

## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage [https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue](https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork23455606-2021-01-01). Save the text of the response as a variable named `html_data`.


In [6]:
tesla_data.reset_index(inplace=True)
tesla_data.head()

Date      Open      High       Low     Close  \
0 2010-06-29 00:00:00-04:00  1.266667  1.666667  1.169333  1.592667   
1 2010-06-30 00:00:00-04:00  1.719333  2.028000  1.553333  1.588667   
2 2010-07-01 00:00:00-04:00  1.666667  1.728000  1.351333  1.464000   
3 2010-07-02 00:00:00-04:00  1.533333  1.540000  1.247333  1.280000   
4 2010-07-06 00:00:00-04:00  1.333333  1.333333  1.055333  1.074000   

      Volume  Dividends  Stock Splits  
0  281494500        0.0           0.0  
1  257806500        0.0           0.0  
2  123282000        0.0           0.0  
3   77097000        0.0           0.0  
4  103003500        0.0           0.0

In [7]:
header = {
    'User-Agent':'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36'
}

Parse the html data using `beautiful_soup`.


In [8]:
url = "https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue"
html_data = requests.get(url, headers = header, timeout = 10).text
soup = BeautifulSoup(html_data, "html5lib")

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Quarterly Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [9]:
tables = soup.find_all('table')
for index,table in enumerate(tables):
  if("Tesla Quarterly Revenue" in str(table)):
    table_index = index
print(table_index)

1


In [10]:
tesla_revenue = pd.DataFrame(columns=['Date','Revenue'])
for row in tables[table_index].tbody.find_all('tr'):
  col = row.find_all('td')
  if(col != []):
    date = col[0].text
    revenue = col[1].text
    d1 = pd.DataFrame({'Date':date, 'Revenue':revenue}, index=[0])
    tesla_revenue = pd.concat([tesla_revenue, d1], ignore_index=True)
tesla_revenue

Date  Revenue
0   2023-12-31  $25,167
1   2023-09-30  $23,350
2   2023-06-30  $24,927
3   2023-03-31  $23,329
4   2022-12-31  $24,318
5   2022-09-30  $21,454
6   2022-06-30  $16,934
7   2022-03-31  $18,756
8   2021-12-31  $17,719
9   2021-09-30  $13,757
10  2021-06-30  $11,958
11  2021-03-31  $10,389
12  2020-12-31  $10,744
13  2020-09-30   $8,771
14  2020-06-30   $6,036
15  2020-03-31   $5,985
16  2019-12-31   $7,384
17  2019-09-30   $6,303
18  2019-06-30   $6,350
19  2019-03-31   $4,541
20  2018-12-31   $7,226
21  2018-09-30   $6,824
22  2018-06-30   $4,002
23  2018-03-31   $3,409
24  2017-12-31   $3,288
25  2017-09-30   $2,985
26  2017-06-30   $2,790
27  2017-03-31   $2,696
28  2016-12-31   $2,285
29  2016-09-30   $2,298
30  2016-06-30   $1,270
31  2016-03-31   $1,147
32  2015-12-31   $1,214
33  2015-09-30     $937
34  2015-06-30     $955
35  2015-03-31     $940
36  2014-12-31     $957
37  2014-09-30     $852
38  2014-06-30     $769
39  2014-03-31     $621
40  2013-12-31     $615
41  2013-09-30     $431
42  2013-06-30     $405
43  2013-03-31     $562
44  2012-12-31     $306
45  2012-09-30      $50
46  2012-06-30      $27
47  2012-03-31      $30
48  2011-12-31      $39
49  2011-09-30      $58
50  2011-06-30      $58
51  2011-03-31      $49
52  2010-12-31      $36
53  2010-09-30      $31
54  2010-06-30      $28
55  2010-03-31      $21
56  2009-12-31         
57  2009-09-30      $46
58  2009-06-30      $27

Execute the following line to remove the comma and dollar sign from the `Revenue` column.


In [11]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace('$',"")
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',',"")
tesla_revenue

Date Revenue
0   2023-12-31   25167
1   2023-09-30   23350
2   2023-06-30   24927
3   2023-03-31   23329
4   2022-12-31   24318
5   2022-09-30   21454
6   2022-06-30   16934
7   2022-03-31   18756
8   2021-12-31   17719
9   2021-09-30   13757
10  2021-06-30   11958
11  2021-03-31   10389
12  2020-12-31   10744
13  2020-09-30    8771
14  2020-06-30    6036
15  2020-03-31    5985
16  2019-12-31    7384
17  2019-09-30    6303
18  2019-06-30    6350
19  2019-03-31    4541
20  2018-12-31    7226
21  2018-09-30    6824
22  2018-06-30    4002
23  2018-03-31    3409
24  2017-12-31    3288
25  2017-09-30    2985
26  2017-06-30    2790
27  2017-03-31    2696
28  2016-12-31    2285
29  2016-09-30    2298
30  2016-06-30    1270
31  2016-03-31    1147
32  2015-12-31    1214
33  2015-09-30     937
34  2015-06-30     955
35  2015-03-31     940
36  2014-12-31     957
37  2014-09-30     852
38  2014-06-30     769
39  2014-03-31     621
40  2013-12-31     615
41  2013-09-30     431
42  2013-06-30     405
43  2013-03-31     562
44  2012-12-31     306
45  2012-09-30      50
46  2012-06-30      27
47  2012-03-31      30
48  2011-12-31      39
49  2011-09-30      58
50  2011-06-30      58
51  2011-03-31      49
52  2010-12-31      36
53  2010-09-30      31
54  2010-06-30      28
55  2010-03-31      21
56  2009-12-31        
57  2009-09-30      46
58  2009-06-30      27

Execute the following lines to remove an null or empty strings in the Revenue column.


In [12]:
tesla_revenue.dropna(inplace=True)
tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]
tesla_revenue

Date Revenue
0   2023-12-31   25167
1   2023-09-30   23350
2   2023-06-30   24927
3   2023-03-31   23329
4   2022-12-31   24318
5   2022-09-30   21454
6   2022-06-30   16934
7   2022-03-31   18756
8   2021-12-31   17719
9   2021-09-30   13757
10  2021-06-30   11958
11  2021-03-31   10389
12  2020-12-31   10744
13  2020-09-30    8771
14  2020-06-30    6036
15  2020-03-31    5985
16  2019-12-31    7384
17  2019-09-30    6303
18  2019-06-30    6350
19  2019-03-31    4541
20  2018-12-31    7226
21  2018-09-30    6824
22  2018-06-30    4002
23  2018-03-31    3409
24  2017-12-31    3288
25  2017-09-30    2985
26  2017-06-30    2790
27  2017-03-31    2696
28  2016-12-31    2285
29  2016-09-30    2298
30  2016-06-30    1270
31  2016-03-31    1147
32  2015-12-31    1214
33  2015-09-30     937
34  2015-06-30     955
35  2015-03-31     940
36  2014-12-31     957
37  2014-09-30     852
38  2014-06-30     769
39  2014-03-31     621
40  2013-12-31     615
41  2013-09-30     431
42  2013-06-30     405
43  2013-03-31     562
44  2012-12-31     306
45  2012-09-30      50
46  2012-06-30      27
47  2012-03-31      30
48  2011-12-31      39
49  2011-09-30      58
50  2011-06-30      58
51  2011-03-31      49
52  2010-12-31      36
53  2010-09-30      31
54  2010-06-30      28
55  2010-03-31      21
57  2009-09-30      46
58  2009-06-30      27

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [13]:
tesla_revenue.tail(5)

Date Revenue
53  2010-09-30      31
54  2010-06-30      28
55  2010-03-31      21
57  2009-09-30      46
58  2009-06-30      27

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [14]:
gamestop = yf.Ticker('GME')

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [15]:
gme_data = gamestop.history(period = 'max')

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [16]:
gme_data.reset_index(inplace=True)
gme_data.head()

Date      Open      High       Low     Close    Volume  \
0 2002-02-13 00:00:00-05:00  1.620128  1.693350  1.603296  1.691666  76216000   
1 2002-02-14 00:00:00-05:00  1.712707  1.716074  1.670626  1.683250  11021600   
2 2002-02-15 00:00:00-05:00  1.683251  1.687459  1.658002  1.674834   8389600   
3 2002-02-19 00:00:00-05:00  1.666418  1.666418  1.578048  1.607504   7410400   
4 2002-02-20 00:00:00-05:00  1.615920  1.662210  1.603296  1.662210   6892800   

   Dividends  Stock Splits  
0        0.0           0.0  
1        0.0           0.0  
2        0.0           0.0  
3        0.0           0.0  
4        0.0           0.0

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage [https://www.macrotrends.net/stocks/charts/GME/gamestop/revenue](https://www.macrotrends.net/stocks/charts/GME/gamestop/revenue?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork23455606-2021-01-01). Save the text of the response as a variable named `html_data`.


In [17]:
url = "https://www.macrotrends.net/stocks/charts/GME/gamestop/revenue"
html_data = requests.get(url, headers = header).text

Parse the html data using `beautiful_soup`.


In [18]:
soup = BeautifulSoup(html_data, "html5lib")

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Quarterly Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [19]:
tables = soup.find_all('table')
for index,table in enumerate(tables):
  if("GameStop Quarterly Revenue" in str(table)):
    table_index = index

gme_revenue = pd.DataFrame(columns=['Date','Revenue'])
for row in tables[table_index].tbody.find_all('tr'):
  col = row.find_all('td')
  if(col != []):
    date = col[0].text
    revenue = col[1].text
    d1 = pd.DataFrame({'Date':date, 'Revenue':revenue}, index=[0])
    gme_revenue = pd.concat([gme_revenue, d1], ignore_index=True)
gme_revenue

Date Revenue
0   2023-10-31  $1,078
1   2023-07-31  $1,164
2   2023-04-30  $1,237
3   2023-01-31  $2,226
4   2022-10-31  $1,186
5   2022-07-31  $1,136
6   2022-04-30  $1,378
7   2022-01-31  $2,254
8   2021-10-31  $1,297
9   2021-07-31  $1,183
10  2021-04-30  $1,277
11  2021-01-31  $2,122
12  2020-10-31  $1,005
13  2020-07-31    $942
14  2020-04-30  $1,021
15  2020-01-31  $2,194
16  2019-10-31  $1,439
17  2019-07-31  $1,286
18  2019-04-30  $1,548
19  2019-01-31  $3,063
20  2018-10-31  $1,935
21  2018-07-31  $1,501
22  2018-04-30  $1,786
23  2018-01-31  $2,825
24  2017-10-31  $1,989
25  2017-07-31  $1,688
26  2017-04-30  $2,046
27  2017-01-31  $2,403
28  2016-10-31  $1,959
29  2016-07-31  $1,632
30  2016-04-30  $1,972
31  2016-01-31  $3,525
32  2015-10-31  $2,016
33  2015-07-31  $1,762
34  2015-04-30  $2,061
35  2015-01-31  $3,476
36  2014-10-31  $2,092
37  2014-07-31  $1,731
38  2014-04-30  $1,996
39  2014-01-31  $3,684
40  2013-10-31  $2,107
41  2013-07-31  $1,384
42  2013-04-30  $1,865
43  2013-01-31  $3,562
44  2012-10-31  $1,773
45  2012-07-31  $1,550
46  2012-04-30  $2,002
47  2012-01-31  $3,579
48  2011-10-31  $1,947
49  2011-07-31  $1,744
50  2011-04-30  $2,281
51  2011-01-31  $3,693
52  2010-10-31  $1,899
53  2010-07-31  $1,799
54  2010-04-30  $2,083
55  2010-01-31  $3,524
56  2009-10-31  $1,835
57  2009-07-31  $1,739
58  2009-04-30  $1,981
59  2009-01-31  $3,492

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [20]:
gme_revenue["Revenue"] = gme_revenue['Revenue'].str.replace('$',"")
gme_revenue["Revenue"] = gme_revenue['Revenue'].str.replace(',',"")
gme_revenue.dropna(inplace=True)
gme_revenue = gme_revenue[gme_revenue['Revenue'] != ""]
gme_revenue.tail(5)

Date Revenue
55  2010-01-31    3524
56  2009-10-31    1835
57  2009-07-31    1739
58  2009-04-30    1981
59  2009-01-31    3492

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


In [21]:
stock_data_specific = tesla_data[tesla_data.Date <= '2021--06-14']
stock_data_specific

Date        Open        High         Low  \
0    2010-06-29 00:00:00-04:00    1.266667    1.666667    1.169333   
1    2010-06-30 00:00:00-04:00    1.719333    2.028000    1.553333   
2    2010-07-01 00:00:00-04:00    1.666667    1.728000    1.351333   
3    2010-07-02 00:00:00-04:00    1.533333    1.540000    1.247333   
4    2010-07-06 00:00:00-04:00    1.333333    1.333333    1.055333   
...                        ...         ...         ...         ...   
2754 2021-06-08 00:00:00-04:00  207.669998  207.696671  198.500000   
2755 2021-06-09 00:00:00-04:00  200.723328  203.929993  199.210007   
2756 2021-06-10 00:00:00-04:00  201.293335  205.529999  200.166672   
2757 2021-06-11 00:00:00-04:00  203.410004  204.186661  200.506668   
2758 2021-06-14 00:00:00-04:00  204.076660  208.496674  203.059998   

           Close     Volume  Dividends  Stock Splits  
0       1.592667  281494500        0.0           0.0  
1       1.588667  257806500        0.0           0.0  
2       1.464000  123282000        0.0           0.0  
3       1.280000   77097000        0.0           0.0  
4       1.074000  103003500        0.0           0.0  
...          ...        ...        ...           ...  
2754  201.196671   78160200        0.0           0.0  
2755  199.593338   49753800        0.0           0.0  
2756  203.373337   71758800        0.0           0.0  
2757  203.296661   48615900        0.0           0.0  
2758  205.896667   61272000        0.0           0.0  

[2759 rows x 8 columns]

In [22]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

In [23]:
make_graph(tesla_data, tesla_revenue, 'Tesla')

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [24]:
make_graph(gme_data, gme_revenue, 'GameStop')

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork23455606-2021-01-01">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
